lib/chef/provisioning/aws_driver/managed_aws.rb

```diff
@@ -0,0 +1,231 @@
+require 'ipaddr'
+
+class Chef
+  module Provisioning
+    module AWSDriver
+      class ManagedAWS
+        #
+        # Create a new ManagedAWS getter.
+        #
+        # @param managed_entries [Chef::Provisioning::ManagedEntryStore] The storage
+        #        where AWS IDs are associated with Chef names.
+        # @param driver [Chef::Provisioning::Driver] The driver from which to get the
+        #        object.
+        def initialize(managed_entries, driver)
+          @managed_entries = managed_entries
+          @driver = driver
+        end
+
+        attr_reader :managed_entries
+        attr_reader :driver
+
+        #
+        # Take a hash of AWS options and look up their IDs, one by one.
+        #
+        # @param options [Hash] Options to pass to an AWS method, with possible Chef names and resources.
+        #
+        # @return [Hash] Options to pass to an AWS method, with real AWS IDs.
+        #
+        def lookup_options(options)
+          result = {}
+          options.each do |name, value|
+            if name.to_s.end_with?('_ids')
+              type = name[0..-5].to_sym
+              result[name] = Array[value].flatten.map { |v| lookup_aws_id!(type, v) || v }
+            elsif name.to_s.end_with?('s')
+              type = name[0..-2].to_sym
+              result[name] = Array[value].flatten.map { |v| lookup_aws_id!(type, v) || v }
+            elsif name.to_s.end_with?('_id')
+              type = name[0..-4].to_sym
+              result[name] = lookup_aws_id!(type, value) || value
+            else
+              result[name] = lookup_aws_id!(name.to_sym, value) || value
+            end
+          end
+          result
+        end
+
+        #
+        # Look up the AWS object ID.
+        #
+        # @param type The type of AWS object to get.
+        # @param id The ID of the object.
+        # @param required `true` if an error should be raised when the object does not
+        #        exist.  The deepest error possible (such as the 404 response) will be
+        #        raised.  If the input value is `nil`, `nil` will be returned rather
+        #        than an error raised.
+        #
+        # @return The AWS object id.  If the ID cannot be found, `nil` is returned.
+        #         The method may return ids even if the actual AWS object does not
+        #         exist (it won't look in AWS in all cases).
+        #
+        def lookup_aws_id(type, id, required: false)
+          return nil if id.nil?
+
+          id = id.name if id.is_a?(Chef::Resource)
+
+          case type
```

Comment:
It feels weird to have this big case statement here - why not make each resource responsible for looking itself up?